_posts/2021-03-15-elastic-training.md

```diff
@@ -0,0 +1,157 @@
+ 
+- A **dynamic** `discover_hosts.sh` composed by MPIJob controller and synchronized to the launcher pod via ConfigMap
+  - MPIJob controller has a podLister, which can be used to list worker pods readily
+  - the controller filters worker pods with `status.phase == Running` and encode the result into the `discover_hosts.sh`
+  - the ConfigMap is updated when `discover_hosts.sh` is modified and the change will be propagated to the launcher pod by Kubernetes
+- A **static** `discover_hosts.sh` using `kubectl` in the launcher pod to list all running worker pods from [APIServer](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/)
+ 
+Scenario 2 changes the delivery image instead of the controller. However, as we cannot limit how frequently users will execute the `discover_hosts.sh` script, it poses a potential threat to the APIServer, especially when the count of worker pods is massive.
+ 
+An fixture to scenario 2 is to replace the `kubectl` with a podLister process, removing extra stress from the APIServer. In this way, we install two processes in launcher pod but lack a proper mechanism to keep the podLister alive. Once the podLister dies, there leaves no elasticity for the training job.
+ 
+Therefore we choose the first scenario and map the `disocver_hosts.sh` under `/etc/mpi/`. We also fixed the other compatibility issues after the worker replica configuration changes. For users choose non-elastic mode, just simply ignore `/etc/mpi/discover_hosts.sh`.
+ 
+Concerns comes to scenario 1 as well. There is a delay between the ConfigMap and what `horovodrun` sees from the `discover_hosts.sh` in the launcher pod. This delay, on one hand, can be tweaked by cluster admin and on the other hand, can be considered as tiny compared to the training elapsed time or the time for Elastic Horovod to handle worker changes.
+ 
+## Demo
+ 
+We present a demo to show how to operate an Elastic Horovod job with MPI Operator. 
+ 
+```bash
+bash-5.0$ kubectl create -f ./tensorflow-mnist-elastic.yaml
+mpijob.kubeflow.org/tensorflow-mnist-elastic 
+createdbash-5.0$ kubectl get po
+NAME    READY   STATUS    RESTARTS  AGE
+tensorflow-mnist-elastic-launcher   1/1     Running   0          14s
+tensorflow-mnist-elastic-worker-0   1/1     Running   0          14s
+tensorflow-mnist-elastic-worker-1   1/1     Running   0          14s
+```
+ 
+The job is created with two workers. After the training begins, we change `MPIJob.Spec.MPIReplicaSpecs["Worker"].Replicas` to **3**, adding another worker. Let's check how the `discover_hosts.sh` changes:
+ 
+```bash
+bash-5.0$ kubectl exec tensorflow-mnist-elastic-launcher -- /etc/mpi/discover_hosts.sh
+tensorflow-mnist-elastic-worker-0:1
+tensorflow-mnist-elastic-worker-1:1
+bash-5.0$ kubectl edit mpijob/tensorflow-mnist-elastic
```

Comment:
Maybe we can use `kubectl patch` to be more explicit here, what do you think ?

Reply:
good idea. I updated the script in the new commit.